🐛 Add several missing `@return` type annotations across the codebase

rsimha opened the pull request:
It turns out there are hundreds of missing `@return` annotations across the AMP codebase, which means closure compiler has incomplete type information during minification.

**PR highlights:**
- Enables the `jsdoc/require-returns` lint rule to detect functions that return a value without specifying its type
- Adds `@return {*}` to all functions that are missing an annotation (with a `TODO` to specify the actual type )
- Specifies the correct return type for all errors detected by `gulp check-types`
    
**Notes:**
- This PR was written with the goal of making as few code changes as possible while satisfying the type checker
- The remaining `@return {*}` annotations that don't result in type errors can be gradually fixed in follow up PRs
- From here on, adding a function definition with a missing `@return` annotation will be flagged by `gulp lint`, and adding a function call where the return type is unspecified will be flagged by `gulp check-types`

rsimha commented:
The majority of changes in this PR are within JSDoc. There are a small number of code changes that needed to be made to satisfy the type checker.

Adding several reviewers for visibility. 
@jeffkaufman; A4A
@lannka: Ads
@zhouyx: Analytics
@erwinmombay: Babel plugins
@danielrozenberg: Infra
@jridgewell: Runtime
@newmuis: Stories
@dvoytenko: Subscriptions
@sparhami: UI

jeffkaufman commented on ads/google/a4a/utils.js:
```diff
@@ -621,24 +621,28 @@ export function getCsiAmpAnalyticsConfig() {
+      'slotId': a4a.element.getAttribute('data-amp-slot-index'),
+      'viewerLastVisibleTime': viewer.getLastVisibleTime() - navStart,
+    }
+  );
```
This change doesn't make sense to me.  How is Object.assign helping?

rsimha replied:
`Object.assign` fixes this invalid cast error:
```js
ads/google/a4a/utils.js:642: ERROR - [JSC_INVALID_CAST] invalid cast - must be a subtype or supertype
from: {
  correlator: number,
  slotId: string,
  viewerLastVisibleTime: number
}
to  : JsonObject
  return /** @type {!JsonObject} */ (vars);
                                    ^^^^^^
```

And casting `vars` to `JsonObject` fixes this inconsistent return type error:
```js
ads/google/a4a/utils.js:642: ERROR - [JSC_TYPE_MISMATCH] inconsistent return type
found   : {
  correlator: number,
  slotId: string,
  viewerLastVisibleTime: number
}
required: JsonObject
  return vars;
         ^^^^
```

jeffkaufman replied:
This seems all wrong to me, though I don't know the right way to do it.  I would expect it to be something like:

```
const vars = JsonObject({
  'correlator': getCorrelator(win, ampdoc),
  'slotId': a4a.element.getAttribute('data-amp-slot-index'),
  'viewerLastVisibleTime': viewer.getLastVisibleTime() - navStart,
});
```

Like, this seems like it shouldn't require any casting at all?

Maybe the `JsonObject` type in build-system/amp.extern.js isn't defined correctly?

rsimha replied:
Entirely possible. Here's the definition:

https://github.com/ampproject/amphtml/blob/6b86000a91ccd59023518db7437fc28f7579a5cb/build-system/amp.extern.js#L57-L65

rsimha replied:
I've opened #23567 for this.

rsimha replied:
That worked. Done.

jeffkaufman commented on ads/runative.js:
```diff
@@ -86,10 +88,11 @@ function getAdContainer(global) {
 function getInitAdScript(global, data) {
   const scriptElement = global.document.createElement('script');
-  const initData = getInitData(data);
+  const initData = /** @type {JsonObject} */ (getInitData(data));
```
instead of casting the output of `getInitData`, specify its return type

rsimha replied:
The cast was originally required to fix this conformance violation:
```js
Type checking failed:
ads/runative.js:97: ERROR - [JSC_CONFORMANCE_VIOLATION] Violation: JSON.stringify must be called with a JsonObject
    `NativeAd(${JSON.stringify(initData)});`
                ^^^^^^^^^^^^^^
```

Fixing the return type of `getInitData` needed a call to `parseJson` to convert `Object` to `JsonObject`, to fix this inconsistent return type error:

```js
Type checking failed:
ads/runative.js:73: ERROR - [JSC_TYPE_MISMATCH] inconsistent return type
found   : Object
required: JsonObject
  return initParams;
         ^^^^^^^^^^
```

jeffkaufman commented on extensions/amp-addthis/0.1/post-message-dispatcher.js:
```diff
@@ -57,12 +57,13 @@ export class PostMessageDispatcher {
 
     if (isObject(data)) {
-      return data;
+      return /** @type {!JsonObject} */ (data);
```
Can you annotate the return type of getData instead?

rsimha replied:
The return type of `getData` is already specified. It inherits the return type of `{?JsonObject|string|undefined}`.

The cast is valid because `isObject(data)` is `true`, and is required because of this inconsistent return type error:

```js
Type checking failed:
extensions/amp-addthis/0.1/post-message-dispatcher.js:66: ERROR - [JSC_TYPE_MISMATCH] inconsistent return type
found   : (JsonObject|null|string|undefined)
required: (JsonObject|null|undefined)
      return data;
             ^^^^
```

jeffkaufman replied:
Is there some sort of `isJsonObject` you could use instead of `isObject` so the compiler can be checking this for us?

rsimha replied:
There isn't.

rsimha replied:
I've opened #23567 for this.

jeffkaufman commented on src/cookies.js:
```diff
@@ -100,14 +100,15 @@ export function setCookie(win, name, value, expirationTime, opt_options) {
 /**
  * Attemp to find the HighestAvailableDomain on
  * @param {!Window} win
+ * @return {?string}
```
Can this return `{string}` so callers don't have to cast?

rsimha replied:
Not possible because `null` is an intended return value.

https://github.com/ampproject/amphtml/blob/569d20852833dd1a47e4296ea6354989aa4017ce/src/cookies.js#L154-L162

jeffkaufman replied:
Can you assert not-null instead?

rsimha replied:
This requires knowledge of the inner workings of this code, to determine whether it should be a `dev()` or `user()` level assert, which is out of scope for this PR.

jeffkaufman commented on extensions/amp-ad-exit/0.1/amp-ad-exit.js:
```diff
@@ -312,11 +312,11 @@ export class AmpAdExit extends AMP.BaseElement {
           0,
           1,
-          createFilter(
+          /** @type {!./filters/filter.Filter} */ (createFilter(
```
Have createFilter specify its return type?

rsimha replied:
Already [specified](https://github.com/rsimha/amphtml/blob/2019-07-26-ReturnAnnotation/extensions/amp-ad-exit/0.1/filters/factory.js#L22-L46). The cast is necessary because `undefined` is a possible return type.

jeffkaufman replied:
Reading `createFilter` it sounds like `undefined` happens if it's given an unknown spec type.  But `amp-ad-exit` is the only extension using `createFilter` and all three of its `createFilter` calls shouldn't result in `undefined`.  Giving an unknown spec to `createFilter` looks like a programming error, so how about modifying `createFilter` to throw in that case, and removing the possibility of it returning `undefined`?

rsimha replied:
Good suggestion. Out of scope for this PR.

rsimha commented:
> I do think addressing the casts soon is valuable, though, because otherwise you can think the type system is protecting you from something when it's not.

Definitely agree. In particularly, I'd love to remove the need for all those casts from `Object` to `JsonObject` across the codebase.

sparhami commented on extensions/amp-ad-exit/0.1/amp-ad-exit.js:
```diff
@@ -312,11 +312,11 @@ export class AmpAdExit extends AMP.BaseElement {
           0,
           1,
-          createFilter(
+          /** @type {!./filters/filter.Filter} */ (createFilter(
```
This is not a safe cast, the function explicitly can return `undefined`. Do not use a type cast to change something nullable (or that can be undefined). This can cause problems down the line. In this case, an item is being added to an array that can be `undefined`. Some piece of code later could try to use the array entry, which at that point would be `undefined`, causing an Error. The original context of it being `undefined` is lost, making the problem hard to track down.

Instead, use `devAssert(createFilter(...))`, which will change the type, but also catch the issue at this point rather than later.

rsimha replied:
Done.

jridgewell replied:
Still not done.

rsimha replied:
https://github.com/ampproject/amphtml/pull/23559/commits/9ba45dfc5e971f8531efb8cd7361712d6d0e5077#diff-63cf316e6647dc9b56b2d7969d48f528

sparhami commented on extensions/amp-addthis/0.1/addthis-utils/rot13.js:
```diff
@@ -24,6 +24,7 @@ const CHAR_Z_UPPER = 90;
  *
  * @param {string} input
  * @private
+ * @return {*} TODO: Specify return type
```
This would ideally be:
```
 * @param {string} input
 * @return {*} TODO: Specify return type
 * @private
```

It doesn't make sense for the visibility modifier to interrupt the function signature types. Not sure if it is too much work to change for this PR in all the places.

rsimha replied:
Good catch. I went ahead and manually self-reviewed and fixed all instances. Done.

rsimha commented on extensions/amp-3d-gltf/0.1/amp-3d-gltf.js:
```diff
@@ -153,10 +153,13 @@ export class Amp3dGltf extends AMP.BaseElement {
   listenGltfViewerMessages_() {
     if (!this.iframe_) {
-      return;
+      return null;
```
Fixed by adding a `devAssert` at the calling site of `listenGltfViewerMessages_()`. Here's why I originally returned `null`: 

```js
Type checking failed:
extensions/amp-3d-gltf/0.1/amp-3d-gltf.js:149: ERROR - [JSC_TYPE_MISMATCH] assignment to property unlistenMessage_ of Amp3dGltf
found   : (Function|null|undefined)
required: (Function|null)
    this.unlistenMessage_ = this.listenGltfViewerMessages_();
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

jridgewell commented on builtins/amp-pixel.js:
```diff
@@ -72,6 +72,7 @@ export class AmpPixel extends BaseElement {
   /**
    * Triggers the signal.
    * @private
+   * @return {*} TODO: Specify return type
```
This can return `{undefined}`, but we'll need to change `return this.triggerPromise_` to just `return`.

rsimha replied:
I'm going to punt on this since it's a code change. See https://github.com/ampproject/amphtml/pull/23559#pullrequestreview-267861042

jridgewell commented on extensions/amp-a4a/0.1/cryptographic-validator.js:
```diff
@@ -27,7 +27,10 @@ export const SIGNATURE_VERIFIER_PROPERTY_NAME =
-  /** @param {!Window} win */
+  /**
+   * @param {!Window} win
+   * @return {*} TODO: Specify return type
```
`!SignatureVerifier`

rsimha replied:
Done.

jridgewell commented on src/base-element.js:
```diff
@@ -1089,6 +1094,7 @@ export class BaseElement {
   /**
    * Declares a child element (or ourselves) as a Layer
    * @param {!Element=} opt_element
+   * @return {*} TODO: Specify return type
```
`undefined`

rsimha replied:
Done.

jridgewell commented on src/cookies.js:
```diff
@@ -100,14 +100,15 @@ export function setCookie(win, name, value, expirationTime, opt_options) {
     domain = opt_options.domain;
   } else if (opt_options && opt_options.highestAvailableDomain) {
-    domain = getHighestAvailableDomain(win);
+    domain = /** @type {string} */ (getHighestAvailableDomain(win));
```
Why are we typecasting here? It can be `null`, too.

rsimha replied:
```js
Type checking failed:
src/cookies.js:105: ERROR - [JSC_TYPE_MISMATCH] actual parameter 5 of trySetCookie$$module$src$cookies does not match formal parameter
found   : (null|string|undefined)
required: (string|undefined)
  trySetCookie(win, name, value, expirationTime, domain);
                                                 ^^^^^^
```

jridgewell commented on src/custom-element.js:
```diff
@@ -1866,7 +1866,10 @@ function createBaseCustomElementClass(win) {
-/** @param {!Element} element */
+/**
+ * @param {!Element} element
+ * @return {*} TODO: Specify return type
```
`boolean`

rsimha replied:
Done.

jridgewell commented on src/document-fetcher.js:
```diff
@@ -60,6 +60,7 @@ export function fetchDocument(win, input, opt_init) {
  * @param {string} input
  * @param {!FetchInitDef} init
  * @private
+ * @return {*} TODO: Specify return type
```
`!Promise<!{response: !Response, xhr: !XMLHttpRequest}>`

rsimha replied:
Done.

jridgewell commented on src/element-service.js:
```diff
@@ -232,6 +232,7 @@ export function isExtensionScriptInNode(ampdoc, extensionId) {
  * @param {HTMLHeadElement|Element|ShadowRoot} head
  * @param {string} extensionId
  * @private
+ * @return {*} TODO: Specify return type
```
`boolean`

jridgewell commented on src/extension-analytics.js:
```diff
@@ -178,6 +178,7 @@ export class CustomEventReporterBuilder {
    * @param {string|!Array<string>} request
+   * @return {*} TODO: Specify return type
    */
   track(eventType, request) {
```
`!CustomEventReporterBuilder`

jridgewell commented on src/extension-analytics.js:
```diff
@@ -202,6 +203,7 @@ export class CustomEventReporterBuilder {
    * Call the #build() method to build and get the CustomEventReporter instance.
    * One CustomEventReporterBuilder instance can only build one reporter, which
    * means #build() should only be called once after all eventType are added.
+   * @return {!Object}
```
`!CustomEventReporter`

jridgewell commented on src/friendly-iframe-embed.js:
```diff
@@ -346,6 +347,7 @@ function mergeHtml(spec) {
 /**
  * Exposes `mergeHtml` for testing purposes.
  * @param {!FriendlyIframeSpec} spec
+ * @return {*} TODO: Specify return type
```
`{string}`

jridgewell commented on src/impression.js:
```diff
@@ -168,6 +168,7 @@ function handleReplaceUrl(win) {
 
 /**
  * @param {string} referrer
+ * @return {*} TODO: Specify return type
```
`boolean`

jridgewell commented on src/loader.js:
```diff
@@ -153,6 +153,7 @@ class LoaderBuilder {
   /**
    * Sets the size of the loader based element's size and a few special cases.
    * @private
+   * @return {*} TODO: Specify return type
```
`undefined`

jridgewell commented on src/pass.js:
```diff
@@ -47,8 +47,13 @@ export class Pass {
-    this.boundPass_ = () => this.pass_();
+    /**
+     * @private
+     * @const {!Function}
```
`function()`

jridgewell commented on src/polyfills/custom-elements.js:
```diff
@@ -727,6 +727,7 @@ function polyfill(win) {
    * You can't use the real HTMLElement constructor, because you can't subclass
    * it without using native classes. So, mock its approximation using
    * createElement.
+   * @return {*} TODO: Specify return type
```
`!Element`

rsimha replied:
```js
src/polyfills/custom-elements.js:730: ERROR - [JSC_UNRECOGNIZED_TYPE_ERROR] Bad type annotation. Unknown type Element
It's possible that a local variable called 'Element' is shadowing the intended global namespace.
   * @return {!Element}
               ^
```

jridgewell commented on src/polyfills/custom-elements.js:
```diff
@@ -782,6 +783,7 @@ function polyfill(win) {
 function wrapHTMLElement(win) {
   const {HTMLElement, Reflect, Object} = win;
   /**
+   * @return {*} TODO: Specify return type
```
`!Element`

jridgewell commented on src/sanitizer.js:
```diff
@@ -276,6 +276,7 @@ export function sanitizeTagsForTripleMustache(html) {
  * Tag policy for handling what is valid html in templates.
  * @param {string} tagName
  * @param {!Array<string>} attribs
+ * @return {*} TODO: Specify return type
```
`?{tagName: string, attribs: !Array<string>}`

jridgewell commented on src/service/layers-impl.js:
```diff
@@ -1341,6 +1341,7 @@ function frameParent(doc) {
  * destroyed (eg, a node inside an iframe that was disconnected from the DOM).
  *
  * @param {!Node} node
+ * @return {*} TODO: Specify return type
```
`boolean`

jridgewell commented on src/service/localization.js:
```diff
@@ -37,6 +37,7 @@ const LANGUAGE_CODE_CHUNK_REGEX = /\w+/gi;
  * @param {!Object<string, !../localized-strings.LocalizedStringBundleDef>} localizedStringBundles
  * @param {!Array<string>} languageCodes
  * @param {!LocalizedStringId} localizedStringId
+ * @return {string}
```
`string|null`

jridgewell commented on src/service/localization.js:
```diff
@@ -141,6 +142,7 @@ export class LocalizationService {
    *     used.  The language is based on the language at that part of the
    *     document.  If unspecified, will use the document-level language, if
    *     one exists, or the default otherwise.
+   * @return {string}
```
`string|null`

rsimha replied:
Also had to add `dev().assertString()` to these calling sites:
```js
Type checking failed:
extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js:703: ERROR - [JSC_TYPE_MISMATCH] actual parameter 2 of AmpStoryAutoAds$$module$extensions$amp_story_auto_ads$0_1$amp_story_auto_ads.prototype.createCtaLayer_ does not match formal parameter
found   : (null|string)
required: string
    return this.createCtaLayer_(adPageElement, ctaText, ctaUrl);
                                               ^^^^^^^

extensions/amp-story/0.1/amp-story-share.js:279: ERROR - [JSC_TYPE_MISMATCH] actual parameter 2 of Toast$$module$extensions$amp_story$0_1$toast.show does not match formal parameter
found   : (null|string)
required: (Node|string)
        Toast.show(this.win, failureString);
                             ^^^^^^^^^^^^^

extensions/amp-story/1.0/amp-story-share.js:361: ERROR - [JSC_TYPE_MISMATCH] actual parameter 2 of Toast$$module$extensions$amp_story$1_0$toast.show does not match formal parameter
found   : (null|string)
required: (Node|string)
        Toast.show(this.storyEl, failureString);
                                 ^^^^^^^^^^^^^
```

jridgewell commented on src/service/template-impl.js:
```diff
@@ -407,6 +407,7 @@ export function installTemplatesService(win) {
  * @param {!Window} win
  * @param {string} type
  * @param {!TemplateClassDef} templateClass
+ * @return {*} TODO: Specify return type
```
`undefined`

jridgewell commented on src/service/url-expander/expander.js:
```diff
@@ -292,6 +292,7 @@ export class Expander {
    *    the FOO binding opt_args will be [[Result of BAR, Result of BAR], [123]].
    *    This structure is so that the outer array will have the correct number of
    *    arguments, but we still can resolve each macro separately.
+   * @return {*} TODO: Specify return type
```
`string|!Promise<string>`

jridgewell commented on src/service/video-manager-impl.js:
```diff
@@ -100,7 +100,11 @@ export class VideoManager {
+    /**
+     * @private
+     * @const
+     * @return {*} TODO: Specify return type
```
`undefined`

jridgewell commented on src/service/vsync-impl.js:
```diff
@@ -452,6 +452,7 @@ export class Vsync {
  * For optimization reasons to stop try/catch from blocking optimization.
  * @param {function(!VsyncStateDef):undefined|undefined} callback
  * @param {!VsyncStateDef} state
+ * @return {*} TODO: Specify return type
```
`boolean`

rsimha commented:
@sparhami I've addressed your comments. Some replies below. PTAL.

> I'm not sure that adding `@return {*} TODO: Specify return type` is a good change to make. It changes something that could be done as a lint (that could be required when changing a file) into a TODO that might never get done.

I don't fully agree with this assessment. There's an advantage to adding all the placeholder `@return {*}` annotations.
- Before this PR, call sites were essentially free to assume the return type.
- Now, `gulp check-types` will detect all instances of unspecified return types. Fixing them will be a one-step process because closure's error messages contain the expected type.
- Not putting in the placeholder and relying on the linter will make it a multi-step fix: Run `gulp lint`, then add a `@return {*}`, then run `gulp check-types`, and then specify the return type.

> I'm not sure if our lint setup allows for it, but i have worked in the past with linting setups that require you to make fixes only when a file is modified as part of a change.

We did this in the past, but abandoned the approach, because it made it unduly onerous to make small changes to files with numerous linter errors that would only show up when the file was edited.

jridgewell commented on src/ssr-template-helper.js:
```diff
@@ -72,7 +72,7 @@ export class SsrTemplateHelper {
    *     the payload. If provided, finding the template in the passed in
    *     element is not attempted.
    * @param {!Object=} opt_attributes Additional JSON to send to viewer.
-   * return {!Promise<{data:{?JsonObject|string|undefined}}>}
```
This doesn't look correct.

rsimha replied:
https://github.com/ampproject/amphtml/pull/23559/commits/9ba45dfc5e971f8531efb8cd7361712d6d0e5077#diff-e0a7330b8f95754a8f79dd9d6eeb035c

jridgewell commented on src/string.js:
```diff
@@ -112,6 +112,7 @@ export function includes(string, substring, start) {
  *   Defaults to 1, but should be set to a larger value your placeholder tokens
  *   can be expanded to other placeholder tokens. Take caution with large values
  *   as recursively expanding a string can be exponentially expensive.
+ * @return {*} TODO: Specify return type
```
`string`

jridgewell commented on src/utils/xhr-utils.js:
```diff
@@ -362,6 +363,7 @@ function normalizeMethod_(method) {
 /**
  * If 415 or in the 5xx range.
  * @param {number} status
+ * @return {*} TODO: Specify return type
```
`boolean`

rsimha commented on extensions/amp-analytics/0.1/config.js:
```diff
@@ -466,6 +468,7 @@ export function mergeObjects(from, to, opt_predefinedConfig) {
  * Expand config's request to object
  * @param {!JsonObject} config
  * @visibleForTesting
+ * @return {?Object}
```
Done.

erwinmombay commented on extensions/amp-3d-gltf/0.1/amp-3d-gltf.js:
```diff
@@ -146,14 +146,17 @@ export class Amp3dGltf extends AMP.BaseElement {
-  /** @private */
+  /**
+   * @private
+   * @return {Function|undefined}
```
`function()|undefined`

rsimha commented on builtins/amp-img.js:
```diff
@@ -211,6 +211,7 @@ export class AmpImg extends BaseElement {
   /**
    * @param {number} newWidth
    * @private
+   * @return {*} TODO: Specify return type
```
Done.

rsimha commented:
> Please file a tracking bug to fix these "TODO: Specify return type" todos.

#23582

rsimha commented:
Tested this locally, Travis is green. Merging before conflicts show up.
